File: mostYearsAlive.js
function mostYearsAlive(arr) {
    const yearAlive = {}
    
    for (let years of arr){
      const yearsRange = createRange(years[0], years[1])
      for (let year of yearsRange){
        yearAlive[year] ? yearAlive[year] += 1 : yearAlive[year] = 1
      }
    }
  
    return findMaxValues(yearAlive).map(Number)
}
  
function createRange(birthYear, deathYear){
   return Array(deathYear - birthYear + 1).fill().map((_, index) => birthYear + index)
}
  
function findMaxValues(yearAliveHash){
    const maxValue = Object.keys(yearAliveHash).reduce((acc, value) => Math.max(acc, yearAliveHash[value]), 0)
    return Object.keys(yearAliveHash).filter(value => yearAliveHash[value] === maxValue)
}

File: most_years_alive.rb
def most_years_alive(arr)
    years_alive_count_hash = Hash.new

    arr.each do |years|
      years_range = Array (years[0]..years[1])
      years_range.each do |year|
        years_alive_count_hash[year]? years_alive_count_hash[year] += 1 : years_alive_count_hash[year] = 1
      end
    end

    return find_max_values(years_alive_count_hash)
end

def find_max_value_keys(years_alive_hash)
   maxAlive = years_alive_hash.values.max
   return years_alive_hash.select { |year, count| count == maxAlive }.keys
end